server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -449,6 +449,29 @@ public Tuple<ShardRouting, ShardRouting> relocateShard(
         return Tuple.tuple(source, target);
     }
 
+    public void relocateOrReinitializeShard(
+        ShardRouting startedShard,
+        String nodeId,
+        long expectedShardSize,
+        RoutingChangesObserver changes
+    ) {
+        if (startedShard.role() == ShardRouting.Role.INDEX_ONLY) {
```

Comment:
I'd prefer we used the `isSearchable` attribute rather than testing the role directly:

```suggestion
        if (startedShard.isSearchable() == false) {
```